### PR TITLE
Using a trait for FormattedString

### PR DESCRIPTION
**Pull request by robertbastian**
Also making the crate nostd

### CODE REVIEW

File: experimental/formatted_string/src/formatted_string.rs [hunk 1]
```diff
@@ -3,16 +3,32 @@
...
 
 use crate::FormattedStringBuilderError;
+use alloc::boxed::Box;
+use alloc::string::String;
+use alloc::{vec, vec::Vec};
+use core::str::from_utf8_unchecked;
 
-/** A FormattedStringBuilder with L levels of type annotations. */
+/// A string with L levels of type annotations
+#[cfg_attr(test, derive(Debug, PartialEq))]
```
Comment (Manishearth) [inline experimental/formatted_string/src/formatted_string.rs:12]: issue: we should unconditionally derive, all the derives work on core

File: experimental/formatted_string/src/formatted_string.rs [hunk 2]
```diff
@@ -59,63 +75,75 @@ fn raise_annotation<F: Copy, const L: usize, const L1: usize>(
...
-        string: LayeredFormattedStringBuilder<F, L1>,
+        string: &LayeredFormattedString<F, L1>,
         field: F,
     ) -> Result<&mut Self, FormattedStringBuilderError> {
         assert_eq!(L - 1, L1);
-        if !self.chars.is_char_boundary(pos) {
-            Err(FormattedStringBuilderError::PositionNotCharBoundary)
+        if pos > self.bytes.len() {
+            Err(FormattedStringBuilderError::IndexOutOfBounds(pos))
+        } else if !unsafe { from_utf8_unchecked(&self.bytes) }.is_char_boundary(pos) {
```
Comment (Manishearth) [inline experimental/formatted_string/src/formatted_string.rs:112]: nit: we should have a safety comment explaining why this is okay at each invocation (even if there's a blanket one above)

File: experimental/formatted_string/src/formatted_string.rs [hunk 3]
```diff
@@ -150,30 +182,38 @@ impl<F: Copy> FormattedStringBuilder<F> {
...
         field: F,
     ) -> Result<&mut FormattedStringBuilder<F>, FormattedStringBuilderError> {
-        if !self.chars.is_char_boundary(pos) {
-            Err(FormattedStringBuilderError::PositionNotCharBoundary)
+        if pos > self.bytes.len() {
+            Err(FormattedStringBuilderError::IndexOutOfBounds(pos))
+        } else if !unsafe { from_utf8_unchecked(&self.bytes) }.is_char_boundary(pos) {
+            Err(FormattedStringBuilderError::PositionNotCharBoundary(
+                pos,
+                unsafe { String::from_utf8_unchecked(self.bytes.clone()) },
```
Comment (Manishearth) [inline experimental/formatted_string/src/formatted_string.rs:190]: nit: in the interest of reducing unsafes, both here and above we should do this as a two-step "construct `str` from `[u8]` with `str::from_utf8_unchecked()`" and when converting to an error just call `.to_owned()` or something.

File: experimental/formatted_string/src/formatted_string.rs [hunk 4]
```diff
@@ -3,16 +3,32 @@
 // (online at: https://github.com/unicode-org/icu4x/blob/main/LICENSE ).
 
 use crate::FormattedStringBuilderError;
+use alloc::boxed::Box;
+use alloc::string::String;
+use alloc::{vec, vec::Vec};
+use core::str::from_utf8_unchecked;
```
Comment (Manishearth) [inline experimental/formatted_string/src/formatted_string.rs:9]: nit: typical style for calling free functions is to import their module (`core::str`) and invoke them with the module name (`str::from_utf8_unchecked()`)

**Review by Manishearth — COMMENTED** [review]
Looks good so far, minor issues

**Comment by robertbastian** [comment]
I added another big commit to dedpulicate some code (str/FormattedString), PTAL

**Comment by Manishearth** [comment]
I think you haven't pushed yet?

File: experimental/formatted_string/src/formatted_string.rs [hunk 5]
```diff
@@ -0,0 +1,251 @@
...
+        // Yay we can return dangling references for singleton types!
+        &[]
+    }
+}
+
+/// A string with L levels of formatting annotations. The name refers
+/// to the fact that the trait FormattedString is also implemented for
+/// &str, but with 0 levels of annotations there isn't much formatting.
+#[derive(Debug, PartialEq)]
+pub struct ActualFormattedString<F: Copy, const L: usize> {
```
Comment (sffc) [inline experimental/formatted_string/src/formatted_string.rs:46]: Suggestion: I think we should have a single FormattedString[Builder] type, rather than having this intermediate one.

Comment (sffc) [reply experimental/formatted_string/src/formatted_string.rs:46]: Potential type names:

- Delete `ActualFormattedString`
- `FormattedStringBuilder` --> `FormattedString`
- `FormattedString` --> `FormattedStringLike`

File: experimental/formatted_string/src/formatted_string.rs [hunk 6]
```diff
@@ -0,0 +1,219 @@
...
+            pos..pos,
+            (0..string.as_ref().len()).map(|i| {
+                let top_level = (
+                    if i == 0 {
+                        LocationInPart::Begin
+                    } else {
+                        LocationInPart::Extend
+                    },
+                    field,
+                );
```
Comment (sffc) [inline experimental/formatted_string/src/formatted_string.rs:132]: Note: I think we likely will want to revise this logic a bit, but let's wait until we have more users like DTF and NF.  For example, when putting together a formatted number, several digits might be appended separately, but they should all coalesce into the same field.

File: experimental/formatted_string/src/formatted_string.rs [hunk 7]
```diff
@@ -0,0 +1,219 @@
...
+    }
+}
+
+impl<F: Copy, const L: usize> FormattedStringLike<F, L> for FormattedString<F, L> {
+    fn annotation_at(&self, pos: usize) -> &[(LocationInPart, F); L] {
+        &self.annotations[pos]
+    }
+}
+
+impl<F: Copy, const L: usize> FormattedString<F, L> {
```
Comment (sffc) [inline experimental/formatted_string/src/formatted_string.rs:61]: Note: I see you went back on what you had originally proposed, which was something like `MultiLevelFormattedString<N>` and `FormattedString` which is an alias for the larger type with N=1.  Why the change of heart?  (I think it's okay either way)

**Review by sffc — APPROVED** [review]

File: experimental/formatted_string/src/formatted_string.rs [hunk 8]
```diff
@@ -0,0 +1,221 @@
...
+            .prepend(&"hello", Field::Word);
+
+        assert_eq!(x.as_ref(), "hello world");
+
+        for i in 0.."hello".len() {
+            assert_eq!(x.field_at(i), Field::Word);
+        }
+        assert_eq!(x.field_at(5), Field::Space);
+        for i in 0.."world".len() {
+            assert_eq!(x.field_at(i), Field::Word);
```
Comment (Manishearth) [inline experimental/formatted_string/src/formatted_string.rs:180]: issue: 

```suggestion
            assert_eq!(x.field_at(6 + i), Field::Word);
```

**Review by Manishearth — COMMENTED** [review]
overall seems good

File: experimental/formatted_string/src/formatted_string.rs (hunk 8 above)
Comment (robertbastian) [reply experimental/formatted_string/src/formatted_string.rs:180]: oh yeah

**Comment by Manishearth** [comment]
(you're probably missing a push?)

**Comment by robertbastian** [comment]
Yep

File: experimental/formatted_string/src/formatted_string.rs [hunk 9]
```diff
@@ -173,7 +175,7 @@ mod test {
         assert_eq!(x.as_ref(), "hello world");
 
         for i in 0.."hello".len() {
-            assert_eq!(x.field_at(i), Field::Word);
+            assert_eq!(x.field_at(6 + i), Field::Word);
```
Comment (sffc) [inline experimental/formatted_string/src/formatted_string.rs:178]: Nit: Please revert

**Review by sffc — APPROVED** [review]

**Review by sffc — APPROVED** [review]
